mypy/expandtype.py

```diff
@@ -29,6 +29,17 @@ def expand_type_by_instance(typ: Type, instance: Instance) -> Type:
         return expand_type(typ, variables)
 
 
+def expand_type_by_instance_bases(typ: Type, instance: Instance) -> Type:
+    if instance.type.tuple_type:
+        # Generic tuple types not supported
+        return typ
+    for base_instance in instance.type.bases:
+        if base_instance.type.bases:
+            typ = expand_type_by_instance_bases(typ, base_instance)
+        typ = expand_type_by_instance(typ, base_instance)
+    return typ
```

Comment:
This algorithm currently looks wrong, and this is why #5664 and similar more complex type variable bindings don't work. You need to map type variables correctly, look for example in `maptype.py`

Reply:
The algorithm is still wrong. For example it will not work for something like this:
```python
class B(Generic[T]):
    @classmethod
    def f(cls) -> T: ...
class C(B[Tuple[T, T]]):
    ...
class D(C[str]):
    ...
reveal_type(D.f())  # Revealed type is 'str', but should be 'Tuple[str, str]'
```
This is not just a small bug there, it is "conceptually" wrong. As I already said twice, you should use something similar to `map_instance_to_supertype` from `maptype.py`. If this is too complex, then I can overtake this PR, it already takes too long.